af/shovel/centrifugation.py

```diff
@@ -326,163 +329,491 @@ def __flush_stubborn(self):
+                    raise RuntimeError('Corrupt index file', autoclaved_index, doc)
+                if autoclaved_xref is not None and report_xref is not None:
+                    write_row(atclv['filename'], t, '{:d}\t0\t{:d}\t{}\t{}\t{}\n'.format(
+                        report_xref,
+                        autoclaved_xref,
+                        pg_quote(doc.get('src_cutoff')), # nullable
+                        pg_quote(rpt['textname']),
+                        pg_binquote(b64decode(rpt['orig_sha1']))))
+                rpt, report_xref = None, None
             elif t == 'frame':
-                frame_off, frame_size, text_off = doc['file_off'], doc['file_size'], doc['text_off']
-            elif t == 'datum' and use_report:
-                intra_off = doc['text_off'] - text_off
-                wbuf.write('{:d}\t{:d}\t{:d}\t{:d}\t{:d}\t\\\\x{}\n'.format(rowno,
-                    frame_off, frame_size, doc['text_off'] - text_off, doc['text_size'],
-                    b64decode(doc['orig_sha1']).encode('hex')))
-            elif t == 'blob':
-                assert False # FIXME: that's some bug, there is no `blob` type AFAIK
-                #yield '{:d}\t\\N\t\\N\t\\N\t\\N\t\\\\x{}\n'.format(rowno,
-                #    b64decode(doc['orig_sha1']).encode('hex'))
-
-def copy_meta_from_index(pgconn, autoclaved_index, bucket):
-    # Writing directly to `report` table is impossible as part of `report`
-    # table is parsed from data files, same is true for `measurement` table.
+                if frm is not None:
+                    raise RuntimeError('Corrupt index file', autoclaved_index, doc)
+                frm = doc # there is `text_size` as well, but that's not that useful
+            elif t == '/frame':
+                if frm is None:
+                    raise RuntimeError('Corrupt index file', autoclaved_index, doc)
+                frm = None
+            elif t == 'datum':
+                if atclv is None or rpt is None:
+                    raise RuntimeError('Corrupt index file', autoclaved_index, doc)
+                if autoclaved_xref is not None and report_xref is not None:
+                    write_row(atclv['filename'], t, '0\t{:d}\t{:d}\t{:d}\t{:d}\t{:d}\t{}\n'.format(
+                        report_xref,
+                        frm['file_off'],
+                        frm['file_size'],
+                        doc['text_off'] - frm['text_off'], # intra-frame off, as text_off is absolute
+                        doc['text_size'], # intra-frame size equals to stream size :)
+                        pg_binquote(b64decode(doc['orig_sha1']))))
+            else:
+                raise RuntimeError('Unknown index.json.gz row type', autoclaved_index, doc)
+    # `dest_wbuf` is not managed with closing() to avoid masking one exception
+    # with another in case of failure.
+    for wbuf in dest_wbuf.itervalues():
+        wbuf.close()
+    with pgconn.cursor() as c:
```

Comment:
This should probably be wrapped in a transaction and probably also placed inside of a separate function that takes the argument `bucket`

Reply:
> This should probably be wrapped in a transaction

`meta_pg()` defines "large" transaction in `with pgconn, ...: # main transaction`.